src/benchmarks/micro/libraries/System.IO/BinaryReaderTests.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using BenchmarkDotNet.Attributes;
+using MicroBenchmarks;
+using System.Linq;
+using System.Text;
+
+namespace System.IO.Tests
+{
+    [BenchmarkCategory(Categories.Libraries)]
+    public class BinaryReaderTests
+    {
+        private BinaryReader _dummyReader;
+
+        private MemoryStream _asciiChar, _nonAsciiChar;
+        private BinaryReader _asciiCharReader, _nonAsciiCharReader;
+        private MemoryStream _smallString, _largeString;
+        private BinaryReader _smallStringReader, _largeStringReader;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            _dummyReader = new BinaryReader(new DummyReadStream());
+            _asciiCharReader = new BinaryReader(_asciiChar = new MemoryStream(new byte[] { (byte)'a' }));
+            _nonAsciiCharReader = new BinaryReader(_nonAsciiChar = new MemoryStream(new byte[] { 0xC3, 0xA0 /* U+00E0 */ }), Encoding.UTF8);
+            _smallStringReader = new BinaryReader(_smallString = CreateMemoryStreamFromString("hello world"), Encoding.UTF8);
+
+            _largeStringReader = new BinaryReader(_largeString = CreateMemoryStreamFromString(new string('a', 512)), Encoding.UTF8);
+
+            static MemoryStream CreateMemoryStreamFromString(string value)
+            {
+                var stream = new MemoryStream();
+                var writer = new BinaryWriter(stream, Encoding.UTF8);
+                writer.Write(value);
+                return stream;
+            }
+        }
+
+        [Benchmark]
+        public BinaryReader DefaultCtor() => new BinaryReader(Stream.Null);
+
+        [Benchmark]
+        public bool ReadBool() => _dummyReader.ReadBoolean();
+
+        [Benchmark]
+        public char ReadAsciiChar()
+        {
+            _asciiChar.Position = 0;
+            return _asciiCharReader.ReadChar();
+        }
+
+        [Benchmark]
+        public char ReadNonAsciiChar()
+        {
+            _nonAsciiChar.Position = 0;
+            return _nonAsciiCharReader.ReadChar();
+        }
+
+        [Benchmark]
+        public ushort ReadUInt16() => _dummyReader.ReadUInt16();
+
+        [Benchmark]
+        public uint ReadUInt32() => _dummyReader.ReadUInt32();
+
+        [Benchmark]
+        public ulong ReadUInt64() => _dummyReader.ReadUInt64();
+
+#if NET5_0_OR_GREATER
+        [Benchmark]
+        public Half ReadHalf() => _dummyReader.ReadHalf();
+#endif
+
+        [Benchmark]
+        public float ReadSingle() => _dummyReader.ReadSingle();
+
+        [Benchmark]
+        public double ReadDouble() => _dummyReader.ReadDouble();
```

Comment:
Is my understanding correct that the dummy reader is going to return `0` for all the benchmarks that return numeric types? If that is the case, it would be better to let it perform more work (to mimic more realistic scenario), by for example reading the values used for `BinaryWriter` benchmarks.

Reply:
Isn't the goal of this microbenchmark to measure only the performance of `BinaryReader`? `DummyReadStream` does the absolute minimum that makes reading succeed. The `BinaryWriter` benchmarks do something similar by discarding the data written.

Reply:
> the reader reads first value, it turns out to be 0, so it returns zero and finishes

I don't understand. Only `ReadString` exhibits this behavior and we don't use `DummyReadStream` there. Reading primitive types from a `BinaryReader` will always read one value no matter the binary data.

Reply:
@teo-tsirpanis please excuse me, for some reason I forgot that it's binary deserialization, so it does not depend on the input size like it does for Utf8 strings etc:

https://github.com/dotnet/runtime/blob/e71faeeeccba29c1f06c7020929cfdadbd32401e/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Unsigned.D.cs#L21-L27